Engine: Better resolve handling for search parameters of type 'reference'

kennethmyhra opened the pull request:
For this to take effect on existing implementations it is necessary to rebuild the index by using the Admin UI that comes with Spark.Web or wire up `IndexRebuildService` in your implementaion.

Se independent commits for details.

Resolves #436
Resolves #521 
Resolves #522 
Resolves #526 

Co-authored-by: @whyfate

kennethmyhra committed:
Engine: Better resolve handling for search parameters of type reference

This enhances our resolve handling for reference types profoundly.

This adds ResourceResolver and makes use of
FhirEvalutionContext.ElementResolver to wire up
ResourceResolver.Resolve() in IndexService.

Co-authored-by: haowang <593301179@qq.com>

kennethmyhra committed:
Engine: Wire up ResourceResolver to our IndexService

Co-authored-by: haowang <593301179@qq.com>

whyfate commented:
I think it is better to put the `ModelInfo.SupportResources` in `IFhirModel`, so that the `Fhir ModelInfo` will not be scattered anywhere. I think the original intention of `IFhirModel` is designed like this.

kennethmyhra commented:
I agree, I did it for the time being to not force a major version bump since we then are breaking binary compatibility by changing the IFhirModel interface. But maybe it's time for that anyways.

whyfate commented:
Oh,I see.
It's time to refactor, but it will take a lot of time.

kennethmyhra committed:
Meta: Update backup image of spark example database

New backup image after better resolve handling for reference types in
IndexService.

kennethmyhra committed:
Documentation: Add note on rebuilding the index

kennethmyhra committed:
Documentation: Reformat MigrateFromv1Tov2.md

kennethmyhra committed:
Meta: Add MigrateFromv1Tov2.md to the docs section in the solution file

kennethmyhra committed:
Engine: Null check arguments passed to IndexService's constructor
